### ISSUE TITLE
Crash when multiple values in config_setting flag_values

### ISSUE THREAD

**Issue by moroten** [bazelbuild/bazel#10146]
### Description of the problem / feature request:

Bazel crashes when `flag_values` in a `config_setting` is set to a comma separated list. See crash log at the end.

Note that `config.string_list()` is not documented, so it was when I tried to find out I ran into this crash.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```
cat > WORKSPACE <<EOF
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "bazel_skylib",
    url = "https://github.com/bazelbuild/bazel-skylib/releases/download/0.9.0/bazel_skylib-0.9.0.tar.gz",
    sha256 = "1dde365491125a3db70731e25658dfdd3bc5dbdfd11b840b3e987ecf043c7ca0",
)
EOF

cat > BUILD.bazel <<EOF
load("@bazel_skylib//rules:common_settings.bzl", "string_list_flag")

string_list_flag(
    name = "strlist",
    build_setting_default = ["foo"],
)

config_setting(
    name = "has_foo_and_or_bar",  # Behaviour is not documented
    flag_values = {
        ":strlist": "foo,bar",
    }
)

filegroup(
    name = "my_filegroup",
    srcs = select({
        ":has_foo_and_or_bar": [],
    }),
)
EOF

# The following will crash. See logs at the bottom.
bazel build my_filegroup
```

### What operating system are you running Bazel on?

Linux

### What's the output of `bazel info release`?

`release 1.1.0`

###  Have you found anything relevant by searching the web?

A quick search for  `config_setting` in GitHub issues did not result in any match.

### Any other information, logs, or outputs that you want to share?

Output from bazel when running `bazel build my_filegroup`:
```
$ bazel build my_filegroup
INFO: Writing tracer profile to '[...]/command.profile.gz'
Internal error thrown during build. Printing stack trace: java.lang.RuntimeException: Unrecoverable error while evaluating node '//:has_foo_and_or_bar BuildConfigurationValue.Key[eaf2712d7c2f2e6aab6dc0e94b6c2de0]
 false' (requested by nodes '//:my_filegroup BuildConfigurationValue.Key[eaf2712d7c2f2e6aab6dc0e94b6c2de0] false')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:531)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:399)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.IllegalArgumentException: expected one element but was: <foo, bar>
        at com.google.common.collect.Iterators.getOnlyElement(Iterators.java:316)
        at com.google.common.collect.Iterables.getOnlyElement(Iterables.java:254)
        at com.google.devtools.build.lib.rules.config.ConfigSetting$UserDefinedFlagMatch.fromAttributeValueAndPrerequisites(ConfigSetting.java:485)
        at com.google.devtools.build.lib.rules.config.ConfigSetting.create(ConfigSetting.java:121)
        at com.google.devtools.build.lib.rules.config.ConfigSetting.create(ConfigSetting.java:79)
        at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createRule(ConfiguredTargetFactory.java:338)
        at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createConfiguredTarget(ConfiguredTargetFactory.java:184)
        at com.google.devtools.build.lib.skyframe.SkyframeBuildView.createConfiguredTarget(SkyframeBuildView.java:734)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.createConfiguredTarget(ConfiguredTargetFunction.java:900)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.compute(ConfiguredTargetFunction.java:345)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:454)
        ... 7 more

INFO: Elapsed time: 0.776s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (3 packages loaded, 6 targets configured)
Internal error thrown during build. Printing stack trace: java.lang.RuntimeException: Unrecoverable error while evaluating node '//:has_foo_and_or_bar BuildConfigurationValue.Key[eaf2712d7c2f2e6aab6dc0e94b6c2de0] false' (requested by nodes '//:my_filegroup BuildConfigurationValue.Key[eaf2712d7c2f2e6aab6dc0e94b6c2de0] false')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:531)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:399)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.IllegalArgumentException: expected one element but was: <foo, bar>
        at com.google.common.collect.Iterators.getOnlyElement(Iterators.java:316)
        at com.google.common.collect.Iterables.getOnlyElement(Iterables.java:254)
        at com.google.devtools.build.lib.rules.config.ConfigSetting$UserDefinedFlagMatch.fromAttributeValueAndPrerequisites(ConfigSetting.java:485)
        at com.google.devtools.build.lib.rules.config.ConfigSetting.create(ConfigSetting.java:121)
        at com.google.devtools.build.lib.rules.config.ConfigSetting.create(ConfigSetting.java:79)
        at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createRule(ConfiguredTargetFactory.java:338)
        at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createConfiguredTarget(ConfiguredTargetFactory.java:184)
        at com.google.devtools.build.lib.skyframe.SkyframeBuildView.createConfiguredTarget(SkyframeBuildView.java:734)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.createConfiguredTarget(ConfiguredTargetFunction.java:900)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.compute(ConfiguredTargetFunction.java:345)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:454)
        ... 7 more
java.lang.RuntimeException: Unrecoverable error while evaluating node '//:has_foo_and_or_bar BuildConfigurationValue.Key[eaf2712d7c2f2e6aab6dc0e94b6c2de0] false' (requested by nodes '//:my_filegroup BuildConfigurationValue.Key[eaf2712d7c2f2e6aab6dc0e94b6c2de0] false')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:531)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:399)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.IllegalArgumentException: expected one element but was: <foo, bar>
        at com.google.common.collect.Iterators.getOnlyElement(Iterators.java:316)
        at com.google.common.collect.Iterables.getOnlyElement(Iterables.java:254)
        at com.google.devtools.build.lib.rules.config.ConfigSetting$UserDefinedFlagMatch.fromAttributeValueAndPrerequisites(ConfigSetting.java:485)
        at com.google.devtools.build.lib.rules.config.ConfigSetting.create(ConfigSetting.java:121)
        at com.google.devtools.build.lib.rules.config.ConfigSetting.create(ConfigSetting.java:79)
        at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createRule(ConfiguredTargetFactory.java:338)
        at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createConfiguredTarget(ConfiguredTargetFactory.java:184)
        at com.google.devtools.build.lib.skyframe.SkyframeBuildView.createConfiguredTarget(SkyframeBuildView.java:734)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.createConfiguredTarget(ConfiguredTargetFunction.java:900)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.compute(ConfiguredTargetFunction.java:345)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:454)
FAILED: Build did NOT complete successfully (3 packages loaded, 6 targets configured)
```

**Comment by gregestren** [comment]
This should obey https://docs.bazel.build/versions/master/be/general.html#config_setting_notes:

> If a flag takes multiple values (like `--copt=-Da --copt=-Db` or a list-typed Starlark flag), `values = { "flag": "a" }` matches if "a" is present anywhere in the actual list.
>
> `values = { "myflag": "a,b" }` works the same way: this matches `--myflag=a --myflag=b`, `--myflag=a --myflag=b --myflag=c`, `--myflag=a,b`, and `--myflag=c,b,a`.
>
> Exact semantics vary between flags. For example, `--copt` doesn't support multiple values in the same instance: `--copt=a,b` produces `["a,b"]` while `--copt=a --copt=b` produces `["a", "b"]` (so `values = { "copt": "a,b" }` matches the former but not the latter).
>
> But `--ios_multi_cpus` (for Apple rules) does: `-ios_multi_cpus=a,b` and `ios_multi_cpus=a --ios_multi_cpus=b` both produce `["a", "b"]`.
>
> Check flag definitions and test your conditions carefully to verify exact expectations.

I acknowledge this is hard to parse. Maybe there's an opportunity for simplification.

**Comment by gregestren** [comment]
P1 (until just a moment ago) issues review: my recollection from my last comment is there are at least two components: the crash and the right design for this sort of thing. The design is more involved and probably involves reviewing how existing multi-value flags work. I'm not prioritizing that from my side at the moment (but anyone would would like to is always welcome!).

So I'm putting this to a P2 and consider the immediate next step to at least stop this from crashing. I think I can get cycles to address that part this week.

**Comment by gregestren** [comment]
I can reproduce the logic error (but not the crash) with `--fat_apk_cpu`. This also uses a command-separated list converter, which means it naturally turns `a,b` into a multi-value list.

Logically, anything that appears in `config_setting` should always be interpreted as a single value.
